llvm/lib/Transforms/Yk/BlockDisambiguate.cpp

```diff
@@ -218,6 +284,32 @@ class YkBlockDisambiguate : public ModulePass {
             BB.replacePhiUsesWith(&BB, DBB);
           }
         }
+      } else if (isa<ReturnInst>(TI)) {
+        // Apply return from self-recursion disambiguation.
+        //
+        // YKFIXME: We do this even if the function not self-recursive.
+        // If we can prove that it is not, then we can skip this step.
+
+        // Make the New Return Block (NRBB) and the Padding Block (PBB).
```

Comment:
You probably could have saved yourself some work here by using [splitBasicBlock](https://llvm.org/doxygen/classllvm_1_1BasicBlock.html#a52c990590792c91dd20b6d45acebe359) (or `splitBasicBlockBefore`), which creates the blocks for you. I haven't tested it but you might just have to call `splitBasicBlock` on the return instruction two times, to generate the same padding. I'll leave it up to you if you want to replace the code, since this this fix is hopefully not going to stay here forever.

Reply:
I'll have a try at using it.

Reply:
Actually, it's not great.

Notice how in the old code, I was holding on to the new unparented blocks until the end before inserting them?

That's because we are in a loop over the blocks of a function, and inserting them immediately would invalidate the iterator.

We can't easily use `splitBasicBlock` for the same reason. It inserts the blocks immediately.

Reply:
Oh I see. That's fair enough.